File: ts/src/lib/budget.ts
/**
 * Reading a transaction's compute budget regardless of its version.
 *
 * Legacy and v0 state their budget in ComputeBudget program instructions, v1
 * in the message config. `computeBudgetOfMessage` reads both from a message
 * that arrived over gRPC, and `computeBudgetOfBase64Transaction` from a signed
 * transaction handed over as base64 — what an RPC response, a facilitator
 * request body, or a wallet's `signTransaction` output carries.
 *
 * Kit's own `getTransactionMessageComputeUnitLimit` does this for messages kit
 * built itself; this module is the equivalent for one that arrived already
 * compiled.
 */

import {
    COMPUTE_BUDGET_PROGRAM_ADDRESS,
    ComputeBudgetInstruction,
    getRequestHeapFrameInstructionDataDecoder,
    getSetComputeUnitLimitInstructionDataDecoder,
    getSetComputeUnitPriceInstructionDataDecoder,
    getSetLoadedAccountsDataSizeLimitInstructionDataDecoder,
    identifyComputeBudgetInstruction,
    MAX_COMPUTE_UNIT_LIMIT,
} from '@solana-program/compute-budget';
import {
    decompileTransactionMessage,
    getAddressDecoder,
    getBase64Encoder,
    getCompiledTransactionMessageDecoder,
    getTransactionDecoder,
    type ReadonlyUint8Array,
    type TransactionVersion,
    type V1TransactionConfig,
} from '@solana/kit';

import { type GrpcMessage, messageVersion } from './grpc';

/** Compute units granted per instruction when a legacy or v0 transaction sets no limit. */
export const DEFAULT_COMPUTE_UNITS_PER_INSTRUCTION = 200_000;

/**
 * A transaction's compute budget, normalised across all three versions.
 *
 * v1's four config fields are the normalisation target. Only
 * `priorityFeeLamports` needs converting: legacy and v0 state a price in
 * micro-lamports per compute unit rather than a total in lamports.
 */
export type ComputeBudget = V1TransactionConfig;

const addressDecoder = getAddressDecoder();

/**
 * Whether a compiled instruction's program account key is the ComputeBudget program.
 *
 * gRPC delivers account keys as raw protobuf `bytes`, so they are decoded to
 * base58 before comparison; any other length cannot be an address.
 */
function isComputeBudgetProgram(accountKey: Uint8Array | undefined): boolean {
    return accountKey?.length === 32 && addressDecoder.decode(accountKey) === COMPUTE_BUDGET_PROGRAM_ADDRESS;
}

/**
 * Renders a budget for display, spelling out what each absent field means.
 *
 * The annotations describe v1's rules: an absent limit resolves to zero, and
 * only `heapSize` falls back to 32 KB. On legacy and v0 an absent compute unit
 * limit resolves to {@link defaultComputeUnitLimit} instead.
 */
export function formatComputeBudget(budget: ComputeBudget, indent = '  '): string {
    const priorityFee =
        budget.priorityFeeLamports === undefined
            ? 'unset (0)'
            : `${budget.priorityFeeLamports} lamports (total, not micro-lamports/CU)`;
    return [
        `${indent}priorityFee:                  ${priorityFee}`,
        `${indent}computeUnitLimit:             ${budget.computeUnitLimit ?? 'unset (0)'}`,
        `${indent}loadedAccountsDataSizeLimit:  ${budget.loadedAccountsDataSizeLimit ?? 'unset (0)'}`,
        `${indent}heapSize:                     ${budget.heapSize ?? 'unset (32KB default)'}`,
    ].join('\n');
}

/** The compute unit limit a legacy or v0 transaction gets when it sets none. */
export function defaultComputeUnitLimit(instructionCount: number): number {
    return Math.min(DEFAULT_COMPUTE_UNITS_PER_INSTRUCTION * instructionCount, MAX_COMPUTE_UNIT_LIMIT);
}

/**
 * Reads the compute budget out of a message received over gRPC.
 *
 * @param message - A transaction message of any version.
 */
export function computeBudgetOfMessage(message: GrpcMessage): ComputeBudget {
    if (messageVersion(message) === 'v1') {
        const config = message.config ?? {};
        return {
            computeUnitLimit: config.computeUnitLimit,
            heapSize: config.heapSize,
            loadedAccountsDataSizeLimit: config.loadedAccountsDataSizeLimit,
            priorityFeeLamports: config.priorityFee === undefined ? undefined : BigInt(config.priorityFee),
        };
    }

    const computeBudgetData = message.instructions
        .filter(instruction => isComputeBudgetProgram(message.accountKeys[instruction.programIdIndex]))
        .map(instruction => instruction.data);

    return computeBudgetOfInstructionData(computeBudgetData, message.instructions.length);
}

/**
 * Reads the compute budget out of a signed transaction encoded as base64.
 *
 * @param base64EncodedTransaction - A signed transaction as it goes on the wire, base64-encoded.
 */
export function computeBudgetOfBase64Transaction(base64EncodedTransaction: string): {
    budget: ComputeBudget;
    version: TransactionVersion;
} {
    const wireTransaction = getBase64Encoder().encode(base64EncodedTransaction);
    const transaction = getTransactionDecoder().decode(wireTransaction);
    const compiled = getCompiledTransactionMessageDecoder().decode(transaction.messageBytes);
    const version: TransactionVersion = compiled.version;

    switch (compiled.version) {
        case 1: {
            const { config } = decompileTransactionMessage(compiled);
            return { budget: { ...config }, version: 1 };
        }
        case 0:
        case 'legacy': {
            const computeBudgetData = compiled.instructions
                .filter(
                    instruction =>
                        compiled.staticAccounts[instruction.programAddressIndex] === COMPUTE_BUDGET_PROGRAM_ADDRESS,
                )
                .map(instruction => instruction.data);
            return {
                budget: computeBudgetOfInstructionData(computeBudgetData, compiled.instructions.length),
                version: compiled.version,
            };
        }
        default: {
            throw new Error(`cannot read a compute budget from transaction version ${String(version)}`);
        }
    }
}

/**
 * Builds a budget out of the data of every ComputeBudget instruction in a
 * legacy or v0 transaction.
 * 
 * @param computeBudgetData - The data of each ComputeBudget instruction, in the order the transaction lists them.
 * @param instructionCount - How many instructions the transaction holds in total, ComputeBudget ones included.
 */
function computeBudgetOfInstructionData(
    computeBudgetData: readonly (ReadonlyUint8Array | undefined)[],
    instructionCount: number,
): ComputeBudget {
    const budget: ComputeBudget = {};
    let priceMicroLamportsPerCu: bigint | undefined;

    for (const data of computeBudgetData) {
        if (data === undefined) {
            continue;
        }
        switch (identifyComputeBudgetInstruction(data)) {
            case ComputeBudgetInstruction.RequestHeapFrame:
                budget.heapSize = getRequestHeapFrameInstructionDataDecoder().decode(data).bytes;
                break;
            case ComputeBudgetInstruction.SetComputeUnitLimit:
                budget.computeUnitLimit = getSetComputeUnitLimitInstructionDataDecoder().decode(data).units;
                break;
            case ComputeBudgetInstruction.SetComputeUnitPrice:
                priceMicroLamportsPerCu = getSetComputeUnitPriceInstructionDataDecoder().decode(data).microLamports;
                break;
            case ComputeBudgetInstruction.SetLoadedAccountsDataSizeLimit:
                budget.loadedAccountsDataSizeLimit =
                    getSetLoadedAccountsDataSizeLimitInstructionDataDecoder().decode(data).accountDataSizeLimit;
                break;
            default:
                break;
        }
    }

    if (priceMicroLamportsPerCu !== undefined) {
        const limit = BigInt(budget.computeUnitLimit ?? defaultComputeUnitLimit(instructionCount));
        // The runtime rounds the total up to whole lamports.
        const microLamports = limit * priceMicroLamportsPerCu;
        budget.priorityFeeLamports = (microLamports + 999_999n) / 1_000_000n;
    }

    return budget;
}

File: rust/src/budget.rs
//! Reading a transaction's compute budget regardless of its version.
//!
//! Legacy and v0 state their budget in ComputeBudget program instructions, v1
//! in the message config. [`ComputeBudget::of_message`] reads both.

use crate::MessageVersion;
use yellowstone_grpc_proto::solana::storage::confirmed_block::Message;

/// The default compute unit limit applied to each instruction when a legacy or
/// v0 transaction does not set one.
pub const DEFAULT_COMPUTE_UNITS_PER_INSTRUCTION: u32 = 200_000;

/// The ceiling a legacy or v0 transaction's implicit compute unit limit is
/// clamped to.
pub const MAX_COMPUTE_UNIT_LIMIT: u32 = 1_400_000;

/// A transaction's compute budget, normalised across all three versions.
#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
pub struct ComputeBudget {
    pub compute_unit_limit: Option<u32>,
    pub heap_size: Option<u32>,
    pub loaded_accounts_data_size_limit: Option<u32>,
    /// The total priority fee in lamports.
    ///
    /// v1 states this directly; legacy and v0 state a price in micro-lamports
    /// per compute unit — see [`ComputeBudget::of_message`].
    pub priority_fee_lamports: Option<u64>,
}

impl ComputeBudget {
    /// Reads the compute budget out of a message received over gRPC.
    ///
    /// v1 reads `config` directly; legacy and v0 scan ComputeBudget program
    /// instructions and convert the fee to a total in lamports so the two are
    /// comparable.
    pub fn of_message(message: &Message) -> Self {
        match crate::grpc::message_version(message) {
            MessageVersion::V1 => Self::from_config(message),
            MessageVersion::Legacy | MessageVersion::V0 => Self::from_instructions(message),
        }
    }

    fn from_config(message: &Message) -> Self {
        let Some(config) = &message.config else {
            return Self::default();
        };
        Self {
            compute_unit_limit: config.compute_unit_limit,
            heap_size: config.heap_size,
            loaded_accounts_data_size_limit: config.loaded_accounts_data_size_limit,
            priority_fee_lamports: config.priority_fee,
        }
    }

    fn from_instructions(message: &Message) -> Self {
        let mut budget = Self::default();
        let mut price_micro_lamports_per_cu = None;

        for instruction in &message.instructions {
            let program_id = message
                .account_keys
                .get(instruction.program_id_index as usize);
            if program_id.map(Vec::as_slice) != Some(solana_sdk_ids::compute_budget::ID.as_array())
            {
                continue;
            }
            let Some((discriminant, operand)) = instruction.data.split_first() else {
                continue;
            };
            match discriminant {
                1 => budget.heap_size = read_u32(operand),
                2 => budget.compute_unit_limit = read_u32(operand),
                3 => price_micro_lamports_per_cu = read_u64(operand),
                4 => budget.loaded_accounts_data_size_limit = read_u32(operand),
                _ => {}
            }
        }

        budget.priority_fee_lamports = price_micro_lamports_per_cu.map(|price| {
            let limit = budget
                .compute_unit_limit
                .unwrap_or_else(|| default_compute_unit_limit(message.instructions.len()));
            // The runtime rounds the total up to whole lamports.
            u64::from(limit).saturating_mul(price).div_ceil(1_000_000)
        });

        budget
    }
}

/// The compute unit limit a legacy or v0 transaction gets when it sets none.
pub fn default_compute_unit_limit(instruction_count: usize) -> u32 {
    let requested = DEFAULT_COMPUTE_UNITS_PER_INSTRUCTION
        .saturating_mul(u32::try_from(instruction_count).unwrap_or(u32::MAX));
    requested.min(MAX_COMPUTE_UNIT_LIMIT)
}

fn read_u32(operand: &[u8]) -> Option<u32> {
    operand.get(..4)?.try_into().ok().map(u32::from_le_bytes)
}

fn read_u64(operand: &[u8]) -> Option<u64> {
    operand.get(..8)?.try_into().ok().map(u64::from_le_bytes)
}
